Toolkit: include a github release utility

srclosson opened the pull request:
**What this PR does / why we need it**: It allows grafana toollkit to perform an automated github publish.

**Special notes for your reviewer**: The pre-commit still is failing. I had to push using `--no-verify`.

srclosson committed:
Adding git publish to toolkit

srclosson committed:
Adding git publish to toolkit

srclosson committed:
grafana-toolkit new feature plugin:github-release

srclosson committed:
grafana-toolkit new feature plugin:github-release

srclosson committed:
Feedback from code review
1. Used async await and exec for executing commands
2. Using dist folder to get plugin information

srclosson committed:
Feedback from code review
1. Used async await and exec for executing commands
2. Using dist folder to get plugin information

srclosson committed:
Reverting change to plugin.json back to original value.

srclosson committed:
Reverting change to plugin.json back to original value.

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=22520) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=22520) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=22520) it.</sub>

ryantxu commented on yarn.lock:
```diff
@@ -2867,6 +2867,11 @@
+  version "6.6.2"
+  resolved "https://registry.yarnpkg.com/@grafana/data/-/data-6.6.2.tgz#34695504cd86bc3b07e4dddcf35ebb1b84158199"
+  integrity sha512-561FlVoE8iFaMB4mo6FBCQc/+J/tv77vzjFWcHE2HB1JwDF15IxbABcDzZWTfPy2ypAi7TR3Kc8Lvss5v0l9Qg==
+
```
I don't think you want any changes to yarn.lock in this PR

srclosson replied:
nooooo! You are absolutely right.

ryantxu commented on packages/grafana-toolkit/src/cli/utils/githubRelease.ts:
```diff
@@ -0,0 +1,65 @@
+class GitHubRelease {
+  token: string;
+  username: string;
+  circleSha1: string;
```
nothing circle specific about it... maybe just `sha1`? or commit hash?

srclosson committed:
reverted changes to yarn.lock

srclosson committed:
reverted changes to yarn.lock

ryantxu commented on packages/grafana-toolkit/src/cli/utils/githubRelease.ts:
```diff
@@ -0,0 +1,65 @@
+      -b "${this.releaseNotes}"
+      -delete
+      "v${getPluginVersion()}"
+      ./artifacts/`);
```
is it really `./artifacts/`? 

The path created by the standard toolkit:ci build is `ci/packages/...`

srclosson committed:
reverted changes to yarn.lock

srclosson committed:
reverted changes to yarn.lock

ryantxu commented on packages/grafana-toolkit/src/cli/utils/execLine.ts:
```diff
@@ -0,0 +1,28 @@
+const distDir = path.resolve(ciDir, 'dist');
+const distContentDir = path.resolve(distDir, getPluginId());
+const pluginJsonFile = path.resolve(distContentDir, 'plugin.json');
+const pluginJson = getPluginJson(pluginJsonFile);
```
weird that this is happening in the root context -- does it really make sense here?

ryantxu replied:
NOTE: this is also why the circle tests are failing

ryantxu commented on packages/grafana-toolkit/src/cli/utils/execLine.ts:
```diff
@@ -0,0 +1,28 @@
+  return pluginJson.info.version;
+};
+
+export { execLine, getPluginVersion, pluginJson };
```
why is execLine exporting getPluginVersion, pluginJson????  weird

ryantxu reviewed (changes requested):
remove changes to yarn.lock and check that the exported funcitons make sense

srclosson committed:
feedback from code review

srclosson committed:
feedback from code review

dprokop commented on packages/grafana-toolkit/src/cli/utils/execLine.ts:
```diff
@@ -0,0 +1,14 @@
+import { exec } from 'child_process';
+
+const execLine = async (command: string): Promise<string> => {
```
Maybe use https://github.com/sindresorhus/execa instead of adding this util? We already use exaca in toolkit

srclosson replied:
Yes :). I can.

ryantxu commented on packages/grafana-toolkit/src/cli/utils/githubRelease.ts:
```diff
@@ -0,0 +1,74 @@
+const GHR_ARCH = process.arch === 'x64' ? 'amd64' : '386';
+const GHR_PLATFORM = ghrPlatform();
+const GHR_EXTENSION = process.platform === 'linux' ? 'tar.gz' : 'zip';
+const PUBLISH_DIR = path.resolve(getCiFolder(), 'packages');
```
Can all this be moved into the beginning of `private getGhr() {`

ryantxu replied:
except for the PUBLISH_DIR ... that can be in release()

srclosson replied:
Roger doger

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/grafana/grafana?pullRequest=22520) <br/>All committers have signed the CLA.

Commit:
feedback from code review 2

Commit:
feedback from code review 2

Commit:
moved constants to recommended functions

Commit:
moved constants to recommended functions

ryantxu commented on packages/grafana-toolkit/src/cli/tasks/plugin.utils.ts:
```diff
@@ -0,0 +1,78 @@
+const execLine = async (line: string): Promise<string> => {
+  const { stdout } = await execa.shell(line);
+  return stdout;
+};
```
is this the same as the promisify stuff?

https://github.com/grafana/grafana/blob/cc638e81f4d24d147941dfaf5e5a47ea1ab41b75/packages/grafana-toolkit/src/cli/tasks/plugin.build.ts#L15

ryantxu replied:
maybe ask @stevenvachon how to best do this -- should not need a special wrapper

srclosson replied:
@dprokop suggested using execa. All this is doing is wrapping it so I didn't have to change all of the calls in the module.

ryantxu commented on packages/grafana-toolkit/src/cli/utils/githubRelease.ts:
```diff
@@ -0,0 +1,71 @@
+  username: string;
+  commitHash: string;
+  releaseNotes: string;
+  constructor(token: string, username: string, commitHash: string, releaseNotes: string) {
```
i like a space between variables and the constructor

Commit:
styling changes and reverting yarn.lock

Commit:
styling changes and reverting yarn.lock

Commit:
removing changes to package.json

Commit:
removing changes to package.json

ryantxu commented on packages/grafana-toolkit/src/cli/tasks/plugin.utils.ts:
```diff
@@ -0,0 +1,78 @@
+  const options = dryrun ? '--dry-run' : '';
+  const GIT_EMAIL = 'eng@grafana.com';
+  const GIT_USERNAME = 'CircleCI Automation';
+  const GITHUB_TOKEN = '';
```
how does this ever get filled in with a real value?   

i assume loaded as an environment variable?

if its missing should we throw an error?

srclosson replied:
good idea. updated.

Commit:
replced execLine with execa

Commit:
replced execLine with execa

Commit:
better error detection around the publish token

Commit:
better error detection around the publish token

ryantxu commented on packages/grafana-toolkit/src/cli/utils/githubRelease.ts:
```diff
@@ -0,0 +1,72 @@
+  release() {
+    const PUBLISH_DIR = path.resolve(getCiFolder(), 'packages');
+    const distDir = path.resolve(process.cwd(), 'dist');
+    const pluginVersion = getPluginJson(path.resolve(distDir, 'plugin.json')).info.version;
```
Maybe use:
```
const pluginInfo = getPluginJson(path.resolve(distDir, 'plugin.json')).info;
const pluginVersion = pluginInfo.version;
const commitHash = pluginInfo.build.hash;
```
this.commitHash -- looks like it is not really used

ryantxu replied:
or if it works fine without it... then we should remove it

srclosson committed:
made simpler with commitHash from build

srclosson committed:
made simpler with commitHash from build

ryantxu reviewed (approved):
💣

srclosson committed:
Testing showed a number of required changes:
- Make the sha configurable
  or through environment variable
  or through git config.
- Allow a release to be recreated
- Set name and repo from git config as this is what
ghr is expecting anyway.
- Appropriate errors if the user
  tries to run a release without
  doing a ci-build and ci-package first.

srclosson committed:
Testing showed a number of required changes:
- Make the sha configurable
  or through environment variable
  or through git config.
- Allow a release to be recreated
- Set name and repo from git config as this is what
ghr is expecting anyway.
- Appropriate errors if the user
  tries to run a release without
  doing a ci-build and ci-package first.

srclosson committed:
Using spinner.
Took out extra dependencies out of project.json
wrote tests manually.

srclosson committed:
Using spinner.
Took out extra dependencies out of project.json
wrote tests manually.

srclosson committed:
Updated tests. Now passing

srclosson committed:
Updated tests. Now passing

srclosson committed:
Merge branch 'toolkit-git-publish' of github.com:grafana/grafana into toolkit-git-publish

srclosson committed:
updated test for reducers, from master

srclosson committed:
package.json and yarn.lock from master
